### PR TITLE
Make SitesList an abstract layer over `state.sites`

### PR DESCRIPTION
**Pull request by Tug**
This is just an experiment to see if we can transform SitesList into an abstract layer over our sites selectors.

### CODE REVIEW

File: client/lib/sites-list/list.js [hunk 1]
```diff
@@ -402,13 +417,16 @@ SitesList.prototype.getSite = function( siteID ) {
...
+	const state = this.getState();
+	let siteFromState;
+
+	if ( Number.isInteger( siteID ) ) {
+		siteFromState = getSite( state, siteID );
+	} else {
+		siteFromState = getSiteBySlug( state, siteID ) || getSiteByDomain( state, siteID ) || getSiteByUrl( state, siteID );
+	}
+
+	return siteFromState && this.createSiteObject( siteFromState );
```
Comment (aduth) [inline client/lib/sites-list/list.js]: I've not looked, but is there duplication of applying `createSiteObject` to the state site, specifically for computed attributes (slug, etc), which are already embedded into the site from `getSite` (vs. `getRawSite`)?

Comment (Tug) [inline client/lib/sites-list/list.js]: Yes, there is ;). This is just a quick draft, we will need to remove duplicate attributes, selectedSite, etc.

**Comment by aduth** [comment]
What would be the plan to commit ourselves to ensure that we don't have two sources of truth for sites data in `sites-list`? Would it be wise to replace the underlying `SiteList.prototype.get` (the raw site data) with a Redux-sourced return value to affect most of the rest of the functions?

**Comment by Tug** [comment]
> What would be the plan to commit ourselves to ensure that we don't have two sources of truth for sites data in sites-list? Would it be wise to replace the underlying SiteList.prototype.get (the raw site data) with a Redux-sourced return value to affect most of the rest of the functions?

Yep I think that would be the idea. every properties currently in `SitesList` should disappear imo:
```
function SitesList() {
	this.initialized = false;
	this.fetched = false;
	this.fetching = false;
	this.selected = null;
	this.data = [];
	...
}
```

**Comment by gwwar** [comment]
Interesting approach here @Tug 👍  Let's see how many usages we can remove this week, before getting too deep into the details on this one. If we're lucky we might not need a interim approach.

As a tangent, porting harder things like libs should be possible now too, I've added an outline of how that might be done in #13154

**Comment by alisterscott** [comment]
Closing as this PR is stale - please reopen if necessary

**Pull request closed without merging**